src/InteractiveWindow/Editor/InteractiveWindow.cs

```diff
@@ -289,97 +282,27 @@ internal void AppendOutput(IEnumerable<string> output)
         /// </summary>
         void IInteractiveWindowOperations.Cancel()
         {
-            ClearInput();
-            UIThread(uiOnly =>
-            {
-                uiOnly.EditorOperations.MoveToEndOfDocument(false);
-                uiOnly.UncommittedInput = null;
-            });
-            _historySearch = null;
+            UIThread(uiOnly => uiOnly.Cancel());
         }
 
-        public void HistoryPrevious(string search = null)
+        void IInteractiveWindowOperations.HistoryPrevious(string search)
         {
-            if (_currentLanguageBuffer == null)
-            {
-                return;
-            }
-
-            var previous = _history.GetPrevious(search);
-            if (previous != null)
-            {
-                if (string.IsNullOrWhiteSpace(search))
-                {
-                    // don't store search as an uncommitted history item
-                    StoreUncommittedInputForHistory();
-                }
-
-                SetActiveCodeToHistory(previous);
-                UIThread(uiOnly => uiOnly.EditorOperations.MoveToEndOfDocument(false));
-            }
+            UIThread(uiOnly => uiOnly.HistoryPrevious(search));
```

Comment:
HistoryPrevious is now asynchronous. In general, should these methods return Tasks so the caller can schedule continuations?